### ISSUE TITLE
"Confirmation token is invalid" all though I'm passing @token to the confirmation email

### ISSUE THREAD

**Issue by ChristofferJoergensen** [plataformatec/devise#4641]
I'm using `devise 3.5.7`for authentication in my Rails app.

I have followed the article [How To: Override confirmations so users can pick their own passwords as part of confirmation activation](https://github.com/plataformatec/devise/wiki/How-To:-Override-confirmations-so-users-can-pick-their-own-passwords-as-part-of-confirmation-activation).

I send a confirmation mail like this:

```ruby
# app/views/devise/mailer/confirmation_instructions.html.haml

= link_to 'Confirm account', confirmation_url(@resource, :confirmation_token => @token)
```

But the `@token` variable is the same as the one stored in the users `confirmation_token` field, and not a hashed version of it. According to [the official blog post about this](http://blog.plataformatec.com.br/2013/08/devise-3-1-now-with-more-secure-defaults/) `@token` should be the right variable to pass.

Any idea how to fix or debug this?

Here are the files that are involved in the signup process:

```ruby
# config/initializers/devise.rb

Devise.setup do |config|
  require 'devise/orm/mongoid'
  config.allow_unconfirmed_access_for = 4.days
  ...
end

# app/controllers/registrations_controller.rb

class RegistrationsController < Devise::RegistrationsController

  def create
    params[:user][:foo] = "foo"

    build_resource(sign_up_params)

    if resource.save
      if resource.active_for_authentication?
        sign_up(resource_name, resource)
        sign_in(resource)
        respond_with resource, :location => after_sign_up_path_for(resource)
      else
        set_flash_message :notice, :"signed_up_but_#{resource.inactive_message}" if is_flashing_format?
        expire_data_after_sign_in!
        respond_with resource, :location => after_inactive_sign_up_path_for(resource)
      end
    else
      clean_up_passwords resource
      respond_with resource
    end
  end

  protected

    def after_inactive_sign_up_path_for(resource)
      new_user_confirmation_path
    end
end

# app/models/user.rb

class User
  include Mongoid::Document

  devise :database_authenticatable, :registerable, :trackable, :rememberable, :confirmable

  ...

  # Confirmable
  field :confirmation_token,              :type => String
  field :confirmed_at,                    :type => Time
  field :confirmation_sent_at,            :type => Time
  field :unconfirmed_email,               :type => String # Only if using reconfirmable

  before_create :build_user

  def password_match?
    self.errors[:password] << "can't be blank" if password.blank?
    self.errors[:password_confirmation] << "can't be blank" if password_confirmation.blank?
    self.errors[:password_confirmation] << "does not match password" if password != password_confirmation
    password == password_confirmation && !password.blank?
  end

  def attempt_set_password(params)
    p = {}
    p[:password] = params[:password]
    p[:password_confirmation] = params[:password_confirmation]
    update_attributes(p)
  end

  def password_required?
    if !persisted?
      false
    else
      !password.nil? || !password_confirmation.nil?
    end
  end

  def has_no_password?
    self.encrypted_password.blank?
  end

  def only_if_unconfirmed
    pending_any_confirmation {yield}
  end

  private

  def build_user
    self.bar = "bar"
  end
end
```

**Comment by toobulkeh** [comment]
Do you have the config option `config.allow_insecure_token_lookup` set to `true` in your config?

**Comment by ChristofferJoergensen** [comment]
Hi @toobulkeh 

Adding `config.allow_insecure_token_lookup = true` seems to raise this error:

```ruby
undefined method `allow_insecure_token_lookup=' for Devise:Module
```

The only place where `confirmation_token` is referenced is in the `confirmations_controller.rb`, as shown below, but I don't reference `@user.confirmation_token` anywhere.

```ruby
  def with_unconfirmed_confirmable
    original_token = params[:confirmation_token]
    confirmation_token = Devise.token_generator.digest(User, :confirmation_token, original_token)
    @confirmable = User.find_or_initialize_with_error_by(:confirmation_token, confirmation_token)
    if !@confirmable.new_record?
      @confirmable.only_if_unconfirmed {yield}
    end
  end

  def do_show
    @confirmation_token = params[:confirmation_token]
    @requires_password = true
    self.resource = @confirmable
    render 'devise/confirmations/show'
  end
```

**Comment by toobulkeh** [comment]
Looks like you're customizing some of these variable names (or this is a much older version of devise I'm not familiar with).

Do you have any place overridden making the mailer call outside of the Devise gem? Can you post that controller and any custom methods it may call?

**Comment by ChristofferJoergensen** [comment]
You're right, I have upgraded my Rails version since I first implemented this feature the first time. At the time it was working. Now I replaced the confirmations_controller in accordance with [the WIKI's instructions](https://github.com/plataformatec/devise/wiki/How-To:-Override-confirmations-so-users-can-pick-their-own-passwords-as-part-of-confirmation-activation#for-rails-34--devise-2x-30-devise-352) and it then works. Thanks for your help :-)
